cmd/agent/main.go

```diff
@@ -485,3 +510,75 @@ func reconcileKubelet(ctx context.Context) (bool, error) {
 	}
 	return false, nil
 }
+
+// getIPAddress gets the machines primary outbound IP address by making a UDP
+// request to a random (non-existent) host and taking the local address
+// used to make the request.
+func getIPAddress() string {
+	conn, err := net.Dial("udp", "1.3.3.7:1")
+	if err != nil {
+		return ""
+	}
+	defer conn.Close()
+	return conn.LocalAddr().(*net.UDPAddr).IP.String()
+}
+
+func KubeletNeedsNewCertificate(headers map[string][]string) error {
+
+	ipAddress := getIPAddress()
+	fileSafeIpAddress := strings.ReplaceAll(ipAddress, ".", "-")
+	certFileLocation := fmt.Sprintf("/etc/kubernetes/ssl/kube-kubelet-%s.pem", fileSafeIpAddress)
+	certKeyFileLocation := fmt.Sprintf("/etc/kubernetes/ssl/kube-kubelet-%s-key.pem", fileSafeIpAddress)
+
+	cert, err := tls.LoadX509KeyPair(certFileLocation, certKeyFileLocation)
+	if err != nil && !strings.Contains(err.Error(), "no such file") {
+		return err
+	}
+
+	if len(cert.Certificate) == 0 {
+		headers[RegenerateKubeletCertificate][0] = "true"
+		return nil
+	}
+
+	parsedCert, err := x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return err
+	}
+
+	// regenerate kubelet cert if the
+	// existing cert is going to expire within 3 days
+	certIsExpiring := false
+	if parsedCert.NotAfter.Sub(time.Now()) < time.Hour*72 {
+		certIsExpiring = true
+	}
+
+	currentHostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	// ensure that the current certificate
+	// contains the current hostname
+	hostnameIncludedInSAN := false
+	for _, name := range parsedCert.DNSNames {
+		if name == currentHostname {
+			hostnameIncludedInSAN = true
+		}
+	}
+
+	ipAddressHasChanged := true
+	for _, ip := range parsedCert.IPAddresses {
+		if ipAddress == ip.String() {
+			ipAddressHasChanged = false
+			break
+		}
+	}
+
+	if !hostnameIncludedInSAN || certIsExpiring || ipAddressHasChanged {
+		headers[RegenerateKubeletCertificate][0] = "true"
+	} else {
+		headers[RegenerateKubeletCertificate][0] = "false"
+	}
+
+	return nil
+}
```

Comment:
```suggestion
func KubeletNeedsNewCertificate() (bool, error) {
  ipAddress := getIPAddress()
  fileSafeIpAddress := strings.ReplaceAll(ipAddress, ".", "-")
  certFileLocation := fmt.Sprintf("/etc/kubernetes/ssl/kube-kubelet-%s.pem", fileSafeIpAddress)
	certKeyFileLocation := fmt.Sprintf("/etc/kubernetes/ssl/kube-kubelet-%s-key.pem", fileSafeIpAddress)
	
	return needsNewCertificate(ipAddress, certFileLocation, certKeyFileLocation, time.Hour*72)
}

func needsNewCertificate(ipAddress, certFileLocation, certKeyFileLocation string, duration time.Duration) (bool, error) {
	cert, err := tls.LoadX509KeyPair(certFileLocation, certKeyFileLocation)
	if err != nil && !strings.Contains(err.Error(), "no such file") {
		return false, err
	}

	if len(cert.Certificate) == 0 {
		return true, nil
	}
	
	parsedCert, err := x509.ParseCertificate(cert.Certificate[0])
	if err != nil {
		return false, err
	}
	
	if certExpired(parsedCert, duration) {
		return true, nil
	}
	
	currentHostname, err := os.Hostname()
	if err != nil {
		return false, err
	}
	
	if hostnameIncludedInSAN(parsedCert, hostname) {
		return true, nil
	}
	
	if ipChanged(parsedCert, ipAddress) {
		return true, nil
	}
	
	return false, nil
}

func certExpired(cert *x509.Certificate, duration time.Duration) (bool, error) {
	if cert.NotAfter.Sub(time.Now()) < duration {
		return true, nil
	}
	
	return false, nil
}

func hostnameIncludedInSAN(cert *x509.Certificate, hostname string) bool {
	// ensure that the current certificate
	// contains the current hostname
	hostnameIncludedInSAN := false
	for _, name := range cert.DNSNames {
		if name == hostname {
			return true
		}
	}
	return false
}

func ipChanged(cert *x509.Certificate, ipAddress string) bool {
	for _, ip := range parsedCert.IPAddresses {
		if ipAddress == ip.String() {
			return false
		}
	}
	return true
}
```
Overall it looks good, I would just do some minor refactoring to make the code more testable. This would make it possible to add both test certs in `testdata`, as well as unit tests that verify simple cases like the ip changing and the hostname. This also assumes a change to split determining whether to add the header, and actually adding it into 2 functions.

Reply:
Alrighty, refactored and tests + test data have been added.